### PR TITLE
feat(request-node): TheGraph as data access

### PR DESCRIPTION
**Pull request by benjlevesque**
- uses TheGraph to get the transactions
- waits for TheGraph indexing after transaction persistance
- classic mode still accessible (by not specifying the graph node URL)

Still to do:
- [x] tests
- [x] README update
- [x] repo with TheGraph Subgraph. Done, see https://github.com/RequestNetwork/storage-subgraph
- [x] updatedBetween filters (probably another PR)
- [ ] better typing for events (probably another PR)

### CODE REVIEW

File: packages/request-node/src/thegraph/TheGraphDataAccess.ts [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+            transaction,
+          },
+        ],
+      },
+    };
+  }
+
+  private deletePending(channelId: string) {
+    delete this.pending[channelId];
+  }
```
Comment (yomarion) [inline packages/request-node/src/thegraph/TheGraphDataAccess.ts:193]: Consider moving next to setPending for clarity

File: packages/request-node/src/thegraph/TheGraphDataAccess.ts [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+  async _getStatus(): Promise<any> {
+    return {
+      lastBlock: await this.graphql.getBlockNumber(),
+      storage: await this.storage._getStatus(),
+    };
+  }
+
+  private setPending(
```
Comment (yomarion) [inline packages/request-node/src/thegraph/TheGraphDataAccess.ts:141]: Consider renaming `addPending` ?

File: packages/request-node/src/thegraph/TheGraphDataAccess.ts [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+      return emptyResult;
+    }
+    const { storageResult, transaction } = this.pending[channelId];
+
+    const { transactions } = await this.graphql.getTransactionsByHash(
+      this.pending[channelId].storageResult.id,
+    );
+
+    // if the pending tx is found, remove its state and fetch the real data
+    if (transactions.length > 0) {
```
Comment (yomarion) [inline packages/request-node/src/thegraph/TheGraphDataAccess.ts:169]: What if for the same request/channelId, there are 2 transactions, one confirmed and one pending?

**Review by yomarion — COMMENTED** [review]
Neat, good job !

? I don't understand the benefits of having the subgraph in a separate repo.

For the tests, don't forget: 1 transaction with more than 1 request, 1 request with many transactions including some pending.

**Comment by benjlevesque** [comment]
> Neat, good job !
> 
> ? I don't understand the benefits of having the subgraph in a separate repo.
> 
> For the tests, don't forget: 1 transaction with more than 1 request, 1 request with many transactions including some pending.

I think we should move all subgraphs to the monorepo later, but for now this will do! 
thanks for the tests tips

File: packages/request-node/src/config.ts [hunk 4]
```diff
@@ -99,6 +99,14 @@ export function getStorageWeb3ProviderUrl(): string {
   );
 }
 
+export function getGraphNodeUrl(): string | undefined {
+  return (
+    argv.graphNodeUrl ||
+    process.env.GRAPH_NODE_URL ||
+    defaultValues.ethereumStorage.ethereum.graphNodeUrl
```
[the remark this answers is not in the document]
Comment (benjlevesque) [reply packages/request-node/src/config.ts:106]: they are now (README), I didn't push that commit yet, my bad

File: packages/request-node/src/thegraph/TheGraphDataAccess.ts (hunk 3 above)
Comment (benjlevesque) [reply packages/request-node/src/thegraph/TheGraphDataAccess.ts:169]: nice catch ;) this actually happens a lot! It's handled in `getTransactionsByChannelId` where we merge pending and not pending

Comment (yomarion) [reply packages/request-node/src/thegraph/TheGraphDataAccess.ts:186]: Did you skip this edge case in the end? Maybe I missed something

File: packages/request-node/src/thegraph/TheGraphDataAccess.ts [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+            state: DataAccessTypes.TransactionState.PENDING,
+            timestamp: Utils.getCurrentTimestampInSecond(),
+            transaction,
+          },
+        ],
+      },
+    };
+  }
+
+  protected createPersistTransactionResult(
```
Comment (yomarion) [inline packages/request-node/src/thegraph/TheGraphDataAccess.ts:208]: Just wording but I don't understand the `Result`

**Review by yomarion — APPROVED** [review]
Seems all good to me, except that I have a doubt for the multiple tx edge case

File: packages/request-node/src/thegraph/TheGraphDataAccess.ts (hunk 3 above)
Comment (benjlevesque) [reply packages/request-node/src/thegraph/TheGraphDataAccess.ts:186]: no, it's already handled in another method, no need to change (I think! feel free to double check)

**Review by alexandre-abrioux — APPROVED** [review]
👌

**Pull request merged**